aiorospy/src/aiorospy/action.py

```diff
@@ -196,42 +234,66 @@ def send_goal(self, goal):
 class AsyncActionServer:
     """ Async wrapper around the action server API. """
 
-    def __init__(self, name, action_spec, coro, loop=None):
+    def __init__(self, name, action_spec, coro, loop=None, ):
         """ Initialize an action server. Incoming goals will be processed via the speficied coroutine. """
         self.name = name
         self._loop = loop if loop is not None else asyncio.get_event_loop()
         self._coro = coro
         self._tasks = {}
 
+        self._exc_q = janus.Queue(loop=loop)
+        self._goal_q = janus.Queue(loop=loop)
+        self._cancel_q = janus.Queue(loop=loop)
+
         self._server = ActionServer(
             name, action_spec, goal_cb=self._goal_cb, cancel_cb=self._cancel_cb, auto_start=False)
 
+    async def start(self):
+        # TODO(pbovbel) actionservers don't stop and cause the program to not terminate
         self._server.start()
+        await asyncio.gather(
+            self._process_goals(),
+            self._process_cancels(),
+            _monitor_exceptions(self._exc_q)
```

Comment:
I haven't done a super clean job of factoring out this _monitor_exceptions pattern. TODO we should implement a mixin helper to handle this, since it'll no doubt be useful in other places outside of actions.

Reply:
Factored out into the ExceptionMonitor